### ISSUE TITLE
Question: Proxying client credentials secret for service accounts

### ISSUE THREAD

**Issue by Geod24** [simonrob/email-oauth2-proxy#271]
Hello @simonrob 
First, thank you for this project, it has been a big timesaver to test some assumptions.

I am trying to migrate some applications that use an SMTP services (STARTTLS / AUTH PLAIN) to pure O365. I wanted a simple setup, which could scale (without incurring costs). What I settled on was creating an application, giving it Outlook `SMTP.SendAsApp` permissions, adding the service principal to outlook and giving it access to the needed mailbox, as well as enabling SMTP on those mailboxes.
Some descriptions of the process:
- https://learn.microsoft.com/en-us/exchange/client-developer/legacy-protocols/how-to-authenticate-an-imap-pop-smtp-application-by-using-oauth
- https://stackoverflow.com/a/76268281

This works quite well because it doesn't lead to the creation of a new account, can be added to an existing mailbox (shared or not), and keep the credentials application specific. Unlike other methods, it doesn't force a 1-to-1 or 1-to-any relationship, and can allow sending but not receiving (or the other way around). The `client_id` and `client_secret` are essentially username / password.

Looking into this repository, I saw that https://github.com/simonrob/email-oauth2-proxy/issues/61 let to the implementation of the CCG grant with `AccessAsApp` (there was no support for SMTP at the time, but that has since changed). However, from everything I can see in the code / this repository, the proxy expect the user's password and there's no way to use the `client_secret`.

In short, I would like to have the following configuration:
```
[foo@contoso.onmicrosoft.com]
token_url = https://login.microsoftonline.com/$TENANT/oauth2/v2.0/token
oauth2_flow = client_credentials
oauth2_scope = https://outlook.office365.com/SMTP.AccessAsApp
client_id = 42424242-4242-4242-4242-424242424242
```

And authenticate with `foo@contoso.onmicrosoft.com` / $CLIENT_SECRET.

**Comment by simonrob** [comment]
The proxy does support the CCG flow as explained [here](https://github.com/simonrob/email-oauth2-proxy/blob/52e729913a53dd05257ee6056670a9aa20906edf/emailproxy.config#L141). Your client ID and client secret values are entered into the proxy's configuration file; your email application authenticates with the username and uses any (consistent) value for the password.

I'm not quite sure what you're asking, but it seems like you're suggesting some form of passing through the client secret as the SMTP password? This is not something the proxy currently does, and I'm not aware of any OAuth 2.0 method where this is the case. The CCG flow is certainly not like this – underneath it is the same as any other method in its handling of the IMAP/POP/SMTP connection, and the only visible difference is that no user interaction is required to retrieve the OAuth 2.0 token.

**Comment by Geod24** [comment]
> I'm not quite sure what you're asking, but it seems like you're suggesting some form of passing through the client secret as the SMTP password?

When the mail client is authenticating with the proxy, yes.

> The CCG flow is certainly not like this – underneath it is the same as any other method in its handling of the IMAP/POP/SMTP connection, and the only visible difference is that no user interaction is required to retrieve the OAuth 2.0 token.

The CCG flow is used to access resources that are owned by the client itself. One doesn't need a user's password for CCG, one needs a way to authenticate the app. As you most likely already know, Microsoft [offers a few options](https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-client-creds-grant-flow#get-a-token), and the most familiar for people will be using a shared secret.

The change I'm suggesting is the following:
If CCG flow and no `client_secret` exists, use the password provided to the proxy as `client_secret`.
This would cover the use case I'm mentioning without impacting other users.

**Comment by Geod24** [comment]
> The change I'm suggesting is the following:

Note that this started as a question. Now that it is turning into a feature request, I would be happy to give a shot at implementing it, provided you see value in the use case.

**Comment by simonrob** [comment]
If you can point me to official documentation showing that this is a standard practice then it can be added. For your own use, it can be implemented simply by making the following change [here](https://github.com/simonrob/email-oauth2-proxy/blob/52e729913a53dd05257ee6056670a9aa20906edf/emailproxy.py#L1121):

```diff
-            del params['client_secret']  # client secret can be optional for O365, but we don't want a None entry
+            params['client_secret'] = password  # assume password is client secret when not otherwise provided
```

**Comment by Geod24** [comment]
Thanks, that is what I was looking for. I also had to do the following:
```diff
--- a/emailproxy.py
+++ b/emailproxy.py
@@ -727,7 +727,7 @@ class OAuth2Helper:

         # note that we don't require permission_url here because it is not needed for the client credentials grant flow,
         # and likewise for client_secret here because it can be optional for Office 365 configurations
-        if not (token_url and oauth2_scope and redirect_uri and client_id):
+        if not (token_url and oauth2_scope and client_id):
             Log.error('Proxy config file entry incomplete for account', username, '- aborting login')
             return (False, '%s: Incomplete config file entry found for account %s - please make sure all required '
                            'fields are added (permission_url, token_url, oauth2_scope, redirect_uri, client_id '
```
But otherwise it works exactly as expected.

> If you can point me to official documentation showing that this is a standard practice then it can be added

I believe [this covers it](https://learn.microsoft.com/en-us/exchange/client-developer/legacy-protocols/how-to-authenticate-an-imap-pop-smtp-application-by-using-oauth#use-client-credentials-grant-flow-to-authenticate-smtp-imap-and-pop-connections).

**Comment by simonrob** [comment]
Thanks for following up – I'm glad this is now working.

Thanks also for the link. However, this is just the standard Microsoft documentation for using OAuth 2.0 in various forms with O365 email. What I was after is something showing using the client secret as the IMAP/POP/SMTP password. However, this might not actually exist given that this proxy is probably the sole use-case for that approach.

On balance I can see why this type of thing might be useful. I think it's something that should be put behind an option, though – for instance, something like `use_password_as_client_credentials_secret` in the proxy's configuration file, and even then only then if there's no `client_secret` value present. I'm also undecided about whether to adjust the `redirect_uri` check – currently this is required by the proxy for every method, even if not used for the less-interactive flows. I'll take a look at whether this is worth changing.

**Comment by Geod24** [comment]
> What I was after is something showing using the client secret as the IMAP/POP/SMTP password. However, this might not actually exist given that this proxy is probably the sole use-case for that approach.

The proxy provide a token to SASL XOAUTH2. That token needs to identify a resource owner that has access to `$EMAIL` for the server to accept our connection. That token is obtained using the Client Credential flow.

Looking at [the OAuth2.0 RFC](https://datatracker.ietf.org/doc/html/rfc6749#section-1.3.4):
> The client credentials (or other forms of client authentication) can be used as an authorization grant when the authorization scope is limited to the protected resources under the control of the client, or to protected resources previously arranged with the authorization server.  Client credentials are used as an authorization grant typically when the client is acting on its own behalf (the client is also the resource owner) or is requesting access to protected resources based on an authorization previously arranged with the authorization server.

Hence with CCG, my interpretation is that there is no difference between resource owner (what we usually calls the user), and client. Next, looking at the [Microsoft documentation on credentials](https://learn.microsoft.com/en-us/entra/identity-platform/quickstart-register-app?tabs=certificate#add-credentials):

> Credentials are used by [confidential client applications](https://learn.microsoft.com/en-us/entra/identity-platform/msal-client-applications) that access a web API. Examples of confidential clients are web apps, other web APIs, or service-type and daemon-type applications. Credentials allow your application to authenticate as itself, requiring no interaction from a user at runtime.

So when using CCG, there is no "user" involved. You can currently supply a bad password to the proxy and it will work, because when going through the authorization grant, we don't authenticate the user, but the application.

On the other hand, when using the ["Authorization Code" grant](https://datatracker.ietf.org/doc/html/rfc6749#section-1.3.1), the is a distinction between client and resource owner:
> The authorization code is obtained by using an authorization server
 as an intermediary between the client and resource owner.  Instead of requesting authorization directly from the resource owner, the client directs the resource owner to an authorization server (via its user-agent as defined in [[RFC2616](https://datatracker.ietf.org/doc/html/rfc2616)]), which in turn directs the resource owner back to the client with the authorization code.

My understanding is that the proxy is primarily targeting the authorization code grant (or similar grants where the resource owner and clients are different entity), not CCG. It is also my understanding that using CCG without configuring the app to have `AccessAsApp` permission would be useless. I went through this repository history to see if there were other users, and found that the author of #88 was using `AccessAsApp`: (https://github.com/lakelse/videos/blob/master/04-install-email-oauth2-proxy-youtube/README.md). I was also looking at Google's `service_account` https://github.com/simonrob/email-oauth2-proxy/commit/4991542f4439532fbcac467a74fdf0f3a7798991 and they seem to be essentially the same as CCG.

Taking the client secret from the password essentially means that in CCG mode, the proxy is not a "full" client, but an OAuth token cache that translate calls. Sorry for the wall of text, I figured this context might be useful to you, or someone else doing research for a similar use case.

The TL;DR is that I agree there won't be any other user than this proxy, hence no documentation.

> I think it's something that should be put behind an option, though

I don't mind either way. I find the current interface a bit odd (since you can provide a bad password on the first connection and it works) but my understanding is centered on my use case so I might be missing something.

**Comment by simonrob** [comment]
Thanks for the additional detail and rationale – I really appreciate this.

You're correct that the main target of the proxy was originally the authorization code flow, and the others were only added later, so the proxy's architecture doesn't necessarily fit as closely with them. This is also why there is what might seem like an odd approach around passwords – when using the authorization code flow this approach makes sense given its interactive nature and the separation between the IMAP/POP/SMTP and OAuth 2.0 aspects. For the CCG flow your suggestion may well be more suitable in many cases, and is clearly a useful addition to the proxy.

The [client_credentials_password branch](https://github.com/simonrob/email-oauth2-proxy/tree/client_credentials_password) has this change. To enable, set the option `use_login_password_as_client_credentials_secret = True` in the proxy's configuration file. Make sure to also remove any existing `client_secret` value for accounts you'd like to use in this way – the proxy will default to secrets provided in the configuration file regardless of this option.

I'd appreciate if you could test this and provide any feedback.

**Comment by Geod24** [comment]
This version doesn't work for me. I get:
> Proxy config file entry incomplete for account X@X.onmicrosoft.com - aborting login

This is the error I encountered here: https://github.com/simonrob/email-oauth2-proxy/issues/271#issuecomment-2312746816
Re-introducing that change makes it work.

**Comment by Geod24** [comment]
As a side note, I was trying to optimize my config file, and having the catch-all fallback make this very easy to configure:
```ini
[@service.contoso.com]
token_url = https://login.microsoftonline.com/abcdef12-3456-789a-bcde-f0123456789ab/oauth2/v2.0/token
oauth2_flow = client_credentials
oauth2_scope = https://outlook.office365.com/.default

[app1@service.contoso.com]
client_id = XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX

[app2@service.contoso.com]
client_id = YYYYYYYY-YYYY-YYYY-YYYY-YYYYYYYYYYYY
```

Only downside I can foresee is that I can't have multiple credentials for the same mailbox, but I doubt it will be an issue in practice.

**Comment by simonrob** [comment]
I originally decided not to make that change since it's easy to just put in a `redirect_uri` value, and this value is indeed mandatory for the authorisation code-driven flow. But since it is not needed for the other supported flows (see documentation for [CCG](https://learn.microsoft.com/en-us/azure/active-directory-b2c/client-credentials-grant-flow?pivots=b2c-user-flow#step-3-obtain-an-access-token), [ROPCG](https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth-ropc#authorization-request) and [Google Service Accounts](https://developers.google.com/identity/protocols/oauth2/service-account#httprest)), and the error message was also itself outdated, I have now made this change in 963ec4b86f781d205e834347e8e22b1145dea386. Could you test again?

**Comment by Geod24** [comment]
> Could you test again?

I am now running your branch and it works as expected.

**Comment by simonrob** [comment]
Thanks for confirming. This is now in the latest release.

**Comment by Geod24** [comment]
Thanks for the help!

**Comment by niklasfink** [comment]
Hi @Geod24, I think this approach is a great addition to the proxy, thanks for bringing it up. I wanted to switch my config to this approach as well, however, I'm having issues with the configuration and the SMTP email dispatch fails with 
`<** 430 4.2.0 STOREDRV; mailbox logon failure; STOREDRV.Submission.Exception:ConnectionFailedTransientException.MapiExceptionLogonFailed; Failed to process message due to a transient exception with message [BeginDiagnosticData]Cannot open mailbox. Server = XXXX281MB3300.DEUP281.PROD.OUTLOOK.COM`

I've set up an example configuration as follows:
```
[SMTP-25]
server_address = smtp-mail.outlook.com
server_port = 587
server_starttls = True

[@verifiedmail.com]
token_url = https://login.microsoftonline.com/TENANTID/oauth2/v2.0/token
oauth2_scope = https://outlook.office365.com/.default
oauth2_flow = client_credentials
client_id = APP_REGISTRATION_CLIENT_ID

[emailproxy]
delete_account_token_on_password_error = False
encrypt_client_secret_on_first_use = False
allow_catch_all_accounts = True
use_login_password_as_client_credentials_secret = True
```

The App Registration in Entra ID has the following API Permissions:
![image](https://github.com/user-attachments/assets/9678517d-5d65-42bd-a4e2-8a67afecc58b)
I think the `full_access_as_app` could even be omitted.

Then, the Service Principal behind the App Registration is added to Exchange Online via:
`New-ServicePrincipal -AppId CLIENT_ID -ObjectId ENTERPRISE_APP_OBJECT_ID`

Afterwards, an ApplicationAccessPolicy should be created via `New-ApplicationAccessPolicy`. But in my case it's neither working with or without an AppAccessPolicy.

Using swaks, this is what I get:
```
swaks --to target@example.com --from my@verifiedmail.com --server local-oauth2-proxy --auth-user "my@verifiedmail.com" --auth-password "APP_REGISTRATION_SECRET"
=== Trying local-oauth2-proxy:25...
=== Connected to local-oauth2-proxy.
<-  220 XXXXX81CA0104.outlook.office365.com Microsoft ESMTP MAIL Service ready at Sat, 9 Nov 2024 17:08:32 +0000 [XXF89BE858]
 -> EHLO X
<-  250-XXXXX81CA0104.outlook.office365.com Hello [IP]
<-  250-SIZE 157286400
<-  250-PIPELINING
<-  250-DSN
<-  250-ENHANCEDSTATUSCODES
<-  250-AUTH PLAIN LOGIN
<-  250-8BITMIME
<-  250-BINARYMIME
<-  250-CHUNKING
<-  250 SMTPUTF8
 -> AUTH LOGIN
<-  334 XX
 -> XXX
<-  334 XXX
 -> XXX==
<-  235 2.7.0 Authentication successful
 -> MAIL FROM:<my@verifiedmail.com>
<-  250 2.1.0 Sender OK
 -> RCPT TO:<target@example.com>
<-  250 2.1.5 Recipient OK
 -> DATA
<-  354 Start mail input; end with <CRLF>.<CRLF>
 -> Date: Sat, 09 Nov 2024 18:08:46 +0100
 -> To: target@example.com
 -> From: my@verifiedmail.com
 -> Subject: test Sat, 09 Nov 2024 18:08:46 +0100
 -> Message-Id: <XX@X>
 -> X-Mailer: swaks v20240103.0 jetmore.org/john/code/swaks/
 ->
 -> This is a test mailing
 ->
 ->
 -> .
<** 430 4.2.0 STOREDRV; mailbox logon failure; STOREDRV.Submission.Exception:ConnectionFailedTransientException.MapiExceptionLogonFailed; Failed to process message due to a transient exception with message [BeginDiagnosticData]Cannot open mailbox. Server = XXXXX1MB3300.DEUP281.PROD.OUTLOOK.COM, mdbGuid = 37a29d52 {... more unreadable data...}[EndDiagnosticData] [Hostname=XXXXX81MB3300.DEUP281.PROD.OUTLOOK.COM]
 -> QUIT
<-  221 2.0.0 Service closing transmission channel
=== Connection closed with remote host.
```

Do you have any idea?

**Comment by simonrob** [comment]
It's hard to give any real help here because the proxy's job is completed successfully when you authenticate (`235 2.7.0 Authentication successful`). I can only assume that there is some unrelated issue with your email setup, especially since it is labelled as a "transient" exception

**Comment by Geod24** [comment]
@niklasfink : Are the mailbox permissions properly set ? I see that you are using catch all accounts, we are not.
I have an internal script to generate the applications / secrets and configure it, so I cannot fully share it, but the gist of it is:
```pwsh
$AppSP = Install-MailUser $AppID $DisplayName
Write-Host "Searching Exchange for existing Service Principal matching AppId $($AppSP.App.AppId)"
$ExchangeSP = Install-MailExchangeServicePrincipal $AppSP.App.AppId $AppSP.App.DisplayName $AppSP.ServicePrincipal.Id

Write-Host "Ensuring SMTP is enabled for mailbox $Email"
Set-CASMailbox -Identity $Email -SmtpClientAuthenticationDisabled $false

Write-Host "Ensuring mailbox permissions are set for the application to use"
# TODO: Switch to send-only permission
# https://learn.microsoft.com/en-us/powershell/module/exchange/add-mailboxpermission?view=exchange-ps#parameters
Add-MailboxPermission -Identity $Email -User $ExchangeSP.Id -AccessRights FullAccess
```

The last part, enabling SMTP and adding mailbox permission, might be what's missing. Note that you can do it from the exchange admin UI.

**Comment by niklasfink** [comment]
Hey, thanks for the support. I was trying to get around the `Add-MailboxPermission`, as it cannot be set on a group and needs to be executed on each Mailbox the Service Principal should have access to. In addition, Microsoft recommends using the "new" Exchange RBAC feature explained here: https://learn.microsoft.com/en-us/exchange/permissions-exo/application-rbac
However, this is just not compatible with this proxy at the moment, as it only supports Graph+EWS, but no SMTP email transfer.

So indeed, adding the MailboxPermission via commandline (I don't see a way how the Service Principal could be added on a Mailbox through the Exchange UI), solved the issue.

@simonrob would you see the integration of MS Graph (instead of SMTP towards smtp-mail.outlook.com) as something which could be integrated into the email-oauth2-proxy? I would say it's future-proof then. Something like:
```
[SMTP-25]
server_address = graph.microsoft.com
ms_graph = True
```

And then something like this would be executed:
https://github.com/ggpwnkthx/Microsoft-Graph-SMTP-Relay/blob/main/app/handlers/microsoft_graph.py#L135

If you think that makes sense, I could try to create a PR for it.

**Comment by simonrob** [comment]
Thanks for following up – I'm glad you managed to resolve this.

Re: MS Graph – I'm not sure how this would fit with (or require) the proxy. The project you've pointed to seems to provide everything you need to use MS Graph over SMTP. What would the proxy add here?

Either way, I think things like this are beyond the scope of the proxy, which is pretty tightly focused around the authentication step. You might be interested in the [plugins branch](https://github.com/simonrob/email-oauth2-proxy/tree/plugins/plugins#email-oauth-20-proxy-plugins), which does allow you to interact with other aspects of IMAP/POP/SMTP. I wouldn't object to a PR that adds this functionality as a plugin, but I don't think it belongs as a core feature.
